### PR TITLE
Introducing DynamicProxyIgnoreTypes.

### LINKED ISSUE #3180: Many controllers in *.HttpApi take too long to load a project
Hi,

I'm facing a problem that when I define many many controllers (around 100 controllers), It takes too long to start project *.HttpApi.Host

How can I solve this issue ?

### PR DESCRIPTION
**Pull request by maliming**
Resolve #3180

```
befor:
2020-03-18 21:04:23.124 +07:00 [INF] Starting web host.
2020-03-18 21:10:08.538 +07:00 [INF] User profile is available.

now:
2020-03-25 18:12:03.433 +08:00 [INF] Starting web host.
2020-03-25 18:12:04.765 +08:00 [INF] User profile is available. 
```

Use `DynamicProxyIgnoreTypes `static class or `Options`?

### CODE REVIEW

**Comment by Y2zz** [comment]
Looking forward to this release

**Comment by Y2zz** [comment]
I have a headache because of this problem for a long time start-up time

File: framework/src/Volo.Abp.Features/Volo/Abp/Features/FeatureInterceptorRegistrar.cs [hunk 1]
```diff
@@ -2,14 +2,15 @@
...
+using Volo.Abp.DynamicProxy;
 
 namespace Volo.Abp.Features
 {
     public static class FeatureInterceptorRegistrar
     {
         public static void RegisterIfNeeded(IOnServiceRegistredContext context)
         {
-            if (ShouldIntercept(context.ImplementationType))
+            if (ShouldIntercept(context.ImplementationType) && !DynamicProxyIgnoreTypes.Contains(context.ImplementationType))
```
Comment (hikalkan) [inline framework/src/Volo.Abp.Features/Volo/Abp/Features/FeatureInterceptorRegistrar.cs]: Move `!DynamicProxyIgnoreTypes.Contains(context.ImplementationType)` into the `ShouldIntercept` method (and also in other classes)

File: framework/src/Volo.Abp.Core/Volo/Abp/DynamicProxy/DynamicProxyIgnoreTypes.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    /// Castle's dynamic proxy class feature will have performance issues for some components, such as the controller of Asp net core MVC.
+    /// For related discussions, see: https://github.com/castleproject/Core/issues/486 https://github.com/abpframework/abp/issues/3180
+    /// The Abp framework may enable interceptors for certain components (UOW, Auditing, Authorization, etc.), which requires dynamic proxy classes, but will cause application performance to decline.
+    /// We need to use other methods for the controller to implement interception, such as middleware or MVC / Page filters.
+    /// So we provide some ignored types to avoid enabling dynamic proxy classes.
+    /// By default it is empty. When you use middleware or filters for these components in your application, you can add these types to the list.
+    /// </summary>
+    public static class DynamicProxyIgnoreTypes
+    {
+        public static List<Type> IgnoreTypes { get; } = new List<Type>();
```
Comment (hikalkan) [inline framework/src/Volo.Abp.Core/Volo/Abp/DynamicProxy/DynamicProxyIgnoreTypes.cs]: Rename to `IgnoredTypes`.

File: framework/src/Volo.Abp.AspNetCore.Mvc/Volo/Abp/AspNetCore/Mvc/AbpAspNetCoreMvcModule.cs [hunk 3]
```diff
@@ -38,6 +40,10 @@ public class AbpAspNetCoreMvcModule : AbpModule
     {
         public override void PreConfigureServices(ServiceConfigurationContext context)
         {
+            DynamicProxyIgnoreTypes.IgnoreTypes.AddIfNotContains(typeof(ControllerBase));
```
Comment (hikalkan) [inline framework/src/Volo.Abp.AspNetCore.Mvc/Volo/Abp/AspNetCore/Mvc/AbpAspNetCoreMvcModule.cs]: Add `DynamicProxyIgnoreTypes.Add(...)` method which internally uses the `IgnoreTypes.AddIfNotContains`.

File: framework/src/Volo.Abp.Core/Volo/Abp/DynamicProxy/DynamicProxyIgnoreTypes.cs (hunk 2 above)
Comment (hikalkan) [reply framework/src/Volo.Abp.Core/Volo/Abp/DynamicProxy/DynamicProxyIgnoreTypes.cs]: Use `HashSet<Type>` for a better `Contains` performance (while you don't use the contains normally)

**Comment by hikalkan** [comment]
In addition to my comments,

When we ignore `PageModel`, it will not support UOW, because ABP doesn't implement action filters for page models. Because it is implemented differently: https://docs.microsoft.com/en-us/aspnet/core/razor-pages/filter?view=aspnetcore-3.1

We overcome this problem to make a method virtual and add [UnitOfWork] manually, for some pages (example: https://github.com/abpframework/abp/blob/dev/modules/account/src/Volo.Abp.Account.Web/Pages/Account/Login.cshtml.cs#L90). 
Please check if this change effects it (I think so). And also check if you can implement https://docs.microsoft.com/en-us/aspnet/core/razor-pages/filter?view=aspnetcore-3.1 for the ABP Framework, for UOW.

**Comment by maliming** [comment]
@hikalkan 

I added four filters for razor page: `AbpExceptionPageFilter`, `AbpAuditPageFilter`, `AbpFeaturePageFilter`, `AbpUowPageFilter` and unit tests for it.

**Comment by hikalkan** [comment]
I have to move this to v2.5 since it is a critic and huge change and needs to have more care on review.

**Pull request merged**